src/models/Visitors.php

```diff
@@ -0,0 +1,344 @@
+use yii\web\Request;
+
+
+use DevGroup\Analytics\models\VisitedPage;
+use DevGroup\Analytics\models\Visitor;
+use DevGroup\Analytics\models\Cities;
+use DevGroup\Analytics\models\Countries;
+
+use DevGroup\Analytics\models\VisitorSession;
+
+class Visitors extends Session
+{
+
+    const KEY_COOKIE = 'visitor';
+
+    const VISITOR_NAME_KEY = 'visitor_id';
+
+    const SESSION_NAME_KEY = 'session_id';
+
+    const USER_IF_AUTO = 'user_status';
+
+
+    public $_statusAddLastVisitor = true;
+
+    public $_statusAddLastSession = true;
+
+    public $_statusAddUser = true;
+
+
+    /**
+     * @return int|mixed|null|string
+     */
+    public function getRefererId()
+    {
+        $referer = $this->getReferrer();
+
+        if ($referer !== null) {
+            return $this->getPageId($referer);
+        } else {
+            return null;
+        }
+    }
+
+    public function addLastVisitor($visitorId)
+    {
+        if ($this->_statusAddLastVisitor && $this->get($this::SESSION_NAME_KEY, false)) {
+            VisitorSession::updateAll([
+                'last_action_at' => date('Y-m-d H:i:s'),
+                'last_visited_page_id' => $this->getPageId(Yii::$app->request->url)
+            ], ['=', 'id', $visitorId]);
+        }
+    }
+
+    public function addLastSession($sessionId)
+    {
+        if ($this->_statusAddLastSession && $this->get($this::VISITOR_NAME_KEY, false)) {
+            // incorrect
+            Visitor::updateAll([
+                'last_activity_at' => date('Y-m-d H:i:s'),
+                'last_activity_visited_page_id' => $this->getPageId(Yii::$app->request->url)
+            ], ['=', 'id', $sessionId]);
+        }
+    }
+
+
+    /**
+     * @return int
+     */
+    private function sessionLifeTime()
+    {
+        return time() + 30;
+    }
+
+    public function run()
+    {
+
+        list($visitorId, $sessionId) = $this->init();
+
+        if ($visitorId !== null && $sessionId !== null) {
+
+//             if visitor first logged. add user id
+            if (
+                $this->_statusAddUser
+                && !$this->get($this::USER_IF_AUTO, false)
+                && Yii::$app->user->getId() !== null
+            ) {
+                Visitor::updateAll([
+                    'user_id' => Yii::$app->user->getId()
+                ], ['=', 'id', $visitorId]);
+
+                $this->set($this::USER_IF_AUTO, true);
+            }
+
+
+            $this->addLastVisitor($visitorId);
+
+            $this->addLastSession($sessionId);
+        }
+
+    }
+
+
+    /**
+     * @return string
+     */
+    public function getIpUser()
+    {
+        return Yii::$app->getRequest()->getUserIP();
+    }
+
+    /**
+     * @param $link
+     */
+    public function getPageId($link)
+    {
+        if ($pageId = Yii::$app->cache->get($link)) {
+            return $pageId;
+        } else {
+            $pageId = $this->getPageIdFromDB($link);
+            Yii::$app->cache->set($link, $pageId, 24 * 60 * 60);
```

Comment:
Damn, what is $link and why you are using it as cache key? Why not to prefix? Where are my tag dependencies?